From f01436b8ce8bd7aba3dd0561678f169f994266fc Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:13:50 -0400
Subject: [powerpc] pseries: dynamically added OF nodes need to call of_node_init

Message-id: <1410545655-205645-602-git-send-email-dzickus@redhat.com>
Patchwork-id: 94453
O-Subject: [RHEL7 PATCH 601/626] powerpc/pseries: dynamically added OF nodes need to call of_node_init
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 97a9a7179aad701ab676e6f29eb90766a1acfde2
Author: Tyrel Datwyler <tyreld@linux.vnet.ibm.com>
Date:   Thu Jul 10 14:50:57 2014 -0400

    powerpc/pseries: dynamically added OF nodes need to call of_node_init

    Commit 75b57ecf9 refactored device tree nodes to use kobjects such that they
    can be exposed via /sysfs. A secondary commit 0829f6d1f furthered this rework
    by moving the kobect initialization logic out of of_node_add into its own
    of_node_init function. The inital commit removed the existing kref_init calls
    in the pseries dlpar code with the assumption kobject initialization would
    occur in of_node_add. The second commit had the side effect of triggering a
    BUG_ON during DLPAR, migration and suspend/resume operations as a result of
    dynamically added nodes being uninitialized.

    This patch fixes this by adding of_node_init calls in place of the previously
    removed kref_init calls.

    Fixes: 0829f6d1f69e ("of: device_node kobject lifecycle fixes")
    Cc: stable@vger.kernel.org
    Signed-off-by: Tyrel Datwyler <tyreld@linux.vnet.ibm.com>
    Acked-by: Nathan Fontenot <nfont@linux.vnet.ibm.com>
    Acked-by: Grant Likely <grant.likely@linaro.org>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/pseries/dlpar.c b/arch/powerpc/platforms/pseries/dlpar.c
index 022b38e..2d0b4d6 100644
--- a/arch/powerpc/platforms/pseries/dlpar.c
+++ b/arch/powerpc/platforms/pseries/dlpar.c
@@ -86,6 +86,7 @@ static struct device_node *dlpar_parse_cc_node(struct cc_workarea *ccwa,
  }
 
  of_node_set_flag(dn, OF_DYNAMIC);
+ of_node_init(dn);
 
  return dn;
 }
diff --git a/arch/powerpc/platforms/pseries/reconfig.c b/arch/powerpc/platforms/pseries/reconfig.c
index 0435bb6..1c0a60d 100644
--- a/arch/powerpc/platforms/pseries/reconfig.c
+++ b/arch/powerpc/platforms/pseries/reconfig.c
@@ -69,6 +69,7 @@ static int pSeries_reconfig_add_node(const char *path, struct property *proplist
 
  np->properties = proplist;
  of_node_set_flag(np, OF_DYNAMIC);
+ of_node_init(np);
 
  np->parent = derive_parent(path);
  if (IS_ERR(np->parent)) {
-- 
1.7.1